File: patches/ox_target/web/js/vice-theme.js
/* -----------------------------------------------------------------------------
   vice_hud theme receiver for ox_target   (added by vice_hud; safe to delete)
   -----------------------------------------------------------------------------
   Receives the theme from client/vice_theme.lua and writes it onto :root as CSS
   custom properties, so /hudtheme recolours the target menu live with nothing
   rebuilt or restarted.

   This is the theme half of ox_lib's vice-glass.js and NOT the tagging half:
   ox_target's page has stable class names, so vice-theme.css targets them
   directly and there is nothing to discover at runtime. See the comment at the
   top of that file.

   ox_target's own main.js switches on `event.data.event`; this listens for
   `event.data.action`, so neither script sees the other's messages and the
   order the two listeners run in does not matter.
   -------------------------------------------------------------------------- */
(function () {
    'use strict';

    var root = document.documentElement;

    var VARS = {
        // The map-panel plate. ox_target is built out of .slot, the
        // material the zone bar and vehicle panel are made of -- not the
        // .glass ox_lib's popups use -- so the glass keys are not read here.
        plate: '--vice-plate',
        plateEdge: '--vice-plate-edge',
        plateTopLit: '--vice-plate-toplit',
        text: '--vice-text',
        textDim: '--vice-text-dim',
        accent: '--vice-accent'
    };

    function applyTheme(t) {
        if (!t || typeof t !== 'object') return;

        for (var key in VARS) {
            if (typeof t[key] === 'string' && t[key]) {
                root.style.setProperty(VARS[key], t[key]);
            }
        }

        if (typeof t.radius === 'number') {
            root.style.setProperty('--vice-radius', t.radius + 'px');
        }

        // Three separate switches because the glass material, the accent
        // colours and the display font are three different opinions, and
        // /hudtheme exposes them as three different rows. Kept identical to
        // ox_lib's attribute names so one stylesheet idiom covers both pages.
        root.setAttribute('data-vice-glass', t.glass === false ? '0' : '1');
        root.setAttribute('data-vice-font', t.font === false ? '0' : '1');
        root.setAttribute('data-vice-accent', '1');
    }

    window.addEventListener('message', function (event) {
        var data = event.data;
        if (!data || data.action !== 'viceTheme') return;
        applyTheme(data.data);
    });
})();

File: patches/ox_lib/resource/interface/client/vice_theme.lua
--[[ ---------------------------------------------------------------------------
     vice_hud theme hook  --  BEGIN  (added by vice_hud; safe to delete)
     ---------------------------------------------------------------------------
     Styles ox_lib's popup layer -- progress bars, context menus, dialogs,
     TextUI, notifications -- to match vice_hud's glass surface.

     Placement is handled separately in notify.lua; this file is only about how
     things LOOK. The two are kept apart because placement is per-notification
     and styling is page-wide.

     How it works: the theme is published as data on a state bag by vice_hud,
     read here, and pushed into ox_lib's own NUI page as CSS custom properties.
     Doing it as variables rather than fixed CSS is what lets /hudtheme recolour
     every popup on the server at runtime with nothing rebuilt or restarted.

     Read from two bags, personal first:
       LocalPlayer.state['vice_hud:theme']   set by /hudtheme on this client
       GlobalState['vice_hud:theme']         set by /themepublish, server-wide

     This file only runs inside ox_lib itself -- SendNUIMessage reaches the
     calling resource's own page, so no other resource could push to it. It is
     picked up by the existing `resource/**/client/*.lua` glob in fxmanifest;
     nothing there needed changing.

     If vice_hud is not running, neither bag exists, nothing is ever sent, and
     web/build/vice-glass.css falls back to ox_lib's stock appearance.

     NOTE: an ox_lib update deletes this file. It also overwrites index.html and
     removes the two lines that load vice-glass.css. Restore both, or popups go
     back to the stock look on their own.
     ------------------------------------------------------------------------ ]]

local function currentTheme()
    local ok, s = pcall(function() return LocalPlayer.state['vice_hud:theme'] end)
    if not ok or type(s) ~= 'table' then s = GlobalState['vice_hud:theme'] end
    if type(s) ~= 'table' then return nil end
    return s
end

--- Pushes the theme into the NUI page. Cheap, and safe to call repeatedly.
local function push()
    local t = currentTheme()
    if not t then return end

    SendNUIMessage({
        action = 'viceTheme',
        data = {
            glass = t.glass ~= false,
            font = t.font ~= false,
            radius = t.radius,
            -- The MAP PANEL plate. web/build/vice-glass.css paints popups with
            -- the .slot recipe now, so without these three a themed popup falls
            -- back to the measured defaults in that file and stops following
            -- /hudtheme's opacity slider.
            plate = t.plate,
            plateEdge = t.plateEdge,
            plateTopLit = t.plateTopLit,
            tint = t.tint,
            rim = t.rim,
            spec = t.spec,
            fill = t.fill,
            fillHi = t.fillHi,
            line = t.line,
            text = t.text,
            textDim = t.textDim,
            accent = t.accent,
            success = t.success,
            error = t.error,
            inform = t.inform,
        }
    })
end

--- Re-push whenever either bag changes, so /hudtheme is live without a reload.
AddStateBagChangeHandler('vice_hud:theme', nil, function()
    push()
end)

CreateThread(function()
    -- The NUI page is not necessarily up the instant this file runs, and a
    -- message sent before it is listening is simply lost. A few early retries
    -- cost nothing and save a "why is it unstyled until I open a menu".
    for _ = 1, 10 do
        Wait(500)
        if currentTheme() then
            push()
            break
        end
    end
end)

-- No explicit refresh event: vice_hud re-sets its local bag on start, which
-- fires the handler above. A second path would be one more thing to keep in
-- step for no benefit.

-- vice_hud theme hook -- END -------------------------------------------------
